### ISSUE TITLE
High queries amount can make miss blocks

### ISSUE THREAD

**Issue by RiccardoM** [tendermint/tendermint#6122]
**Tendermint version** (use `tendermint version` or `git rev-parse --verify HEAD` if installed from source):
```
tendermint: ""
abci: 0.17.0
blockprotocol: 11
p2pprotocol: 8
```

**ABCI app** (name for built-in, URL for self-written if it's publicly available):
[Desmos `v0.15.1`](https://github.com/desmos-labs/desmos/tree/v0.15.1)

**Environment**:
- **OS** (e.g. from /etc/os-release):  
   ```
   NAME="Ubuntu"
   VERSION="20.04.1 LTS (Focal Fossa)"
   ID=ubuntu
   ID_LIKE=debian
   PRETTY_NAME="Ubuntu 20.04.1 LTS"
   VERSION_ID="20.04"
   HOME_URL="https://www.ubuntu.com/"
   SUPPORT_URL="https://help.ubuntu.com/"
   BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
   PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
   VERSION_CODENAME=focal
   UBUNTU_CODENAME=focal
   ```
- **Install tools**:
- **Others**:

**What happened**:
> Premise: this issue is a duplicate of cosmos/cosmos-sdk#8602 that I opened here after @marbar3778 suggested to do so. 

We are currently developing [BDJuno](https://github.com/forbole/bdjuno), a tool that allows to listen to a chain state and parses the data into a PostgreSQL database. In order to do so, it acts in two ways at the same time: 

1. Listens for new blocks
2. Parses all old blocks 

For each block, it then reads the different modules' states and stores them inside the PostgreSQL database. What we do is we a snapshot of the state for each block and store it. To do so, we use gRPC to get all the data that can change from one block to another (i.e. delegations, unbonding delegations, redelegations, staking commissions, etc). 

As we also need to parse old blocks and get the state at very old heights, we setup an archive node with `pruning = "nothing"`. 

When we first started our parser, everything was working properly. The node was able to keep up with syncing new blocks and answering to gRPC calls properly. 

Recently, however, we noticed that the node started to lack behind the chain state, was over 500 blocks behind. So, we stopped the parser and let the node catch up again with the chain state. Then, we restarted the parser. One week later and the node is once again more than 1,000 blocks behind the current chain height. 

**Note**  
I have no idea if this happens only because the pruning is set to `nothing`. However, I believe this should be investigated as it might result in some tools (eg. explorers) making the nodes stop in the future if too many requests are done to them. It could even be exploited via a DDoS attack to validator nodes if this results to happen also to nodes that have the pruning option set to `default` or `everything`. 

**What you expected to happen**:  
The node should continue to be in sync with the chain, while responding to queries without starting to lack behind. 

**Have you tried the latest version**:   
No

**How to reproduce it** (as minimally and precisely as possible):  
1. Start a full node with `pruning = "nothing"`
2. Start performing a lot of gRPC requests (around 100 per block) 
3. The node will start to slowly lack behind in block syncing 

**Logs (paste a small part showing an error (< 10 lines) or link a pastebin, gist, etc. containing more of the log file)**:  
N.A.

**Config (you can paste only the changes you've made)**:  
```toml
pruning = "nothing"
```

**Comment by RiccardoM** [comment]
Reporting here some @marbar3778's comments: 

[First](https://github.com/cosmos/cosmos-sdk/issues/8602#issuecomment-780407377): 

> I believe this is a tendermint issue. The RPC is blocking and causes consensus to slow down. This is a known issue and why we recommend validators not expose their rpc to the public network.

[Second](https://github.com/cosmos/cosmos-sdk/issues/8602#issuecomment-780409386)

> All requests in the sdk requests are routed through tendermint. The request goes through the abci_query abci method.

**Comment by RiccardoM** [comment]
I'm not sure, but this might be also partially correlated to #6048

**Comment by alexanderbez** [comment]
@RiccardoM you claimed that once you stopped Juno, the node quickly resumed back to an in-sync state, correct? You turned on Juno again and saw the same issue.

In order to eliminate `pruning=nothing` hypothesis, have you tried to have this node run indefinitely w/o Juno at all? That would tell you if its the pruning or not. If not, then, yes, it could be an issue in RPC mutex contention or something else in the RPC layer. But even so, why does this reflect a slowdown in syncing/p2p gossip?

**Comment by marbar3778** [comment]
> If not, then, yes, it could be an issue in RPC mutex contention or something else in the RPC layer. But even so, why does this reflect a slowdown in syncing/p2p gossip @marbar3778?

The mutex contention bleeds into consensus making it block if enough requests come in. I can reproduce this pretty easily with a two node setup locally running tm-load-test or a tool to request data hundreds of times a second.

**Comment by RiccardoM** [comment]
> @RiccardoM you claimed that once you stopped Juno, the node quickly resumed back to an in-sync state, correct? You turned on Juno again and saw the same issue.

This is correct

> In order to eliminate `pruning=nothing` hypothesis, have you tried to have this node run indefinitely w/o Juno at all? 

Yes, I've been running this node for 2 months without Juno querying it, and it had no problem. It always stayed in sync with the chain. The problem only popped up when I started running Juno

**Comment by alexanderbez** [comment]
Ok so it's not pruning related then.

**Comment by marbar3778** [comment]
> Yes, I've been running this node for 2 months without Juno querying it, and it had no problem. It always stayed in sync with the chain. The problem only popped up when I started running Juno

Did the amount of requests increase?

**Comment by RiccardoM** [comment]
> > Yes, I've been running this node for 2 months without Juno querying it, and it had no problem. It always stayed in sync with the chain. The problem only popped up when I started running Juno
> 
> Did the amount of requests increase?

The amounts of requests increased when I started Juno (which was the only one making the requests). Before that the node was not getting queried at all

**Comment by kwunyeung** [comment]
This issue https://github.com/tendermint/tendermint/issues/3211 becomes related now. We have to keep restarting the explorer node to keep it in sync because of the high query amount but it might result in lost of tx index.

**Comment by carameleon** [comment]
`/cosmos/staking/v1beta1/validators` query's pagination make block sync slow down,

One of our node used by mobile wallet service, slow down periodically.
(Probably, slow down 300+- blocks per hour behind of latest block)

At the moment, we remove the pagination code in our wallet service and then block sync normal (at least now)

**Comment by kwunyeung** [comment]
@carameleon not sure if this related but I did use pagination with `/cosmos/staking/v1beta1/validators` setting 2000 validators per page so that I only query it once.

**Comment by carameleon** [comment]
@kwunyeung 
Yeap, we have applied the same our mobile wallet. In the past, we query with params(100 per page).

In the case of Mintscan, by the way, the node which is connected to Mintscan seems to stop while about 10mins periodically.

**Comment by alexanderbez** [comment]
Is this perhaps then related to the slowdowns operators have been seeing due to gRPC requests being non performant? Are we sure indexing is the issue @kwunyeung?

**Comment by kwunyeung** [comment]
@alexanderbez I didn't mean indexing is the issue causing this. As the way we make the explorer to keep up in sync is to restart the node periodically due to the slow down from gRPC requests, this also could cause the tx indexing missing in the node. We lost a couple tx indexes from the node already. So to me, these two issues are related. If the slow down issue can't be solved, we would get lost tx indexes.

**Comment by alexanderbez** [comment]
I see! Ok, so it's the fact that the gRPC queries are causing significant slowdowns. If that is the case, isn't this the responsibility of the SDK?

**Comment by RiccardoM** [comment]
> I see! Ok, so it's the fact that the gRPC queries are causing significant slowdowns. If that is the case, isn't this the responsibility of the SDK?

Actually, as @marbar3778 already stated, all SDK gRPC queries are redirected to become Tendermint ABCI query. This, associated with #6048, might be why the node slows down when there are a lot of requests. 

See [here](https://github.com/tendermint/tendermint/issues/6122#issuecomment-780453520) as well

**Comment by marbar3778** [comment]
This is both a tendermint and cosmos-sdk issue. 

**Tendermint:**
- fix the mutex contention that leads to missing blocks and consensus falling behind. 

**Cosmos-SDK**
- Don't use tendermint as a query mutex. I'm not sure what needs to happen here but something that comes to mind is concurrent access of state in the sdk.

**Comment by robert-zaremba** [comment]
Is this related to the recent change where we route all queries through ABCI?

**Comment by alexanderbez** [comment]
> Is this related to the recent change where we route all queries through ABCI?

Sounds like it! If you can avoid using Tendermint as a proxy for queries and instead directly query stores for gRPC queries, you should see significant improvements.

TBH, I don't know why `Info` and `Query` ABCI method calls are serialized, but regardless, the SDK should consider skipping Tendermint for gRPC queries./

**Comment by robert-zaremba** [comment]
We can't, we changed it intentionally for security reasons: https://github.com/cosmos/cosmos-sdk/pull/8549

**Comment by robert-zaremba** [comment]
However the queries should be executed in parallel - and we should definitely do that: https://github.com/cosmos/cosmos-sdk/issues/8591

**Comment by carameleon** [comment]
I've investigated our APIs, which are using gRPC. 
As a result, the query `staking/validatordelegations` is the key to node slowdowns.
Just only several calls for `staking/validatordelegations` make the node slow.
Node and Mintscan performance got recovered by disabling this query.

**Comment by kwunyeung** [comment]
@carameleon thanks! That helps.

**Comment by begetan** [comment]
This MR can fix the Tendermint concurrency issue: https://github.com/tendermint/tm-db/pull/239

**Comment by faddat** [comment]
@begetan -- I think that it helps, but perhaps does not fully fix it.
